### PR TITLE
Searchbar hide when scroll

### PR DESCRIPTION
**Pull request by whitzhu**
## Summary

Toolbar that consist of SEARCH & BREADCRUMB
Toolbar hide/show based on user's scrolling activity
Breadcrumb have been refactored so it shows "<- Back" as default on content page and on mobile

**Navbar Hidden**

screen shot 2016-08-10 at 3 37 23 pm

**Navbar Shown**
screen shot 2016-08-10 at 3 37 27 pm

**Updated Breadcrumb**
![screen shot 2016-08-10 at 3 32 42 pm]
(https://cloud.githubusercontent.com/assets/12800136/17573558/1c986b24-5f10-11e6-9caa-3c2e8a2204d3.png)

**Updated Breadcrumb:  "<- Back" for all content page via Explore**
screen shot 2016-08-10 at 3 32 49 pm

**Updated Breadcrumb:  "<- Back" for all conent page via Explore on Mobile**
screen shot 2016-08-10 at 3 32 57 pm

**Updated Breadcrumb:  "<- Back" for all explore pages via Mobile**
screen shot 2016-08-10 at 3 33 01 pm

**currnet view of Search Panel on mobile**
screen shot 2016-08-10 at 3 33 06 pm

**Temporarily moved search input and results down so it's not covered by toolbar. This will be revisited in another PR**
screen shot 2016-08-10 at 3 25 22 pm

**moved channel to toolbar. Addded mobile styling to channel**
96fc7a12-5fe3-11e6-8b57-45d0424ab1b0

NOTE: The search panel needs some work, will re-visit. 
NOTE: Toolbar might be covering content header, need to adjust content position in another PR

*Short description*

## TODO

- [ ] Have tests been written for the new code?
- [ ] Has documentation been written/updated?
- [ ] New dependencies (if any) added to requirements file
- [ ] Add an entry to CHANGELOG.rst
- [ ] Add yourself it AUTHORS.rst if you don't appear there

## Reviewer guidance

*If you PR has a significant size, give the reviewer some helpful remarks*

## Issues addressed

List the issues solved or partly solved by the PR

## Documentation

If the PR has documentation, link the file here (either .rst in your repo or if built on Read The Docs)

## Screenshots (if appropriate)

They're nice. :)

### CODE REVIEW

File: kolibri/plugins/learn/assets/src/vue/toolbar/index.vue [hunk 1]
```diff
@@ -0,0 +1,95 @@
+<template>
+
+  <div v-bind:class="['toolbar-show', displayToolBar ? 'toolbar-hide' : '' ]" >
+    <breadcrumbs class="breadcrumbs"></breadcrumbs>
+    <search-button v-on:scrolling="handleScroll" class='search-btn'></search-button>
```
Comment (indirectlylit) [inline kolibri/plugins/learn/assets/src/vue/toolbar/index.vue]: let's use the `@scrolling` shorthand

File: kolibri/plugins/learn/assets/src/vue/breadcrumbs/index.vue [hunk 2]
```diff
@@ -1,38 +1,92 @@
...
+      <span class="parent full-breadcrumbs" v-if="pageMode === $options.PageModes.EXPLORE && pageName !== $options.PageNames.EXPLORE_CONTENT && !isRoot">
+        <a v-link="exploreRoot">Explore</a>
+      </span>
+      <span v-if="pageName === $options.PageNames.LEARN_CONTENT">
+        <a v-link="learnRoot">
+          <span class='sep'>
+            <svg height="24" viewbox="0 0 24 24" width="24" xmlns="http://www.w3.org/2000/svg">
+              <path d="M0 0h24v24H0z" fill="none"></path>
+              <path d="M20 11H7.83l5.59-5.59L12 4l-8 8 8 8 1.41-1.41L7.83 13H20v-2z"></path>
+            </svg>
```
Comment (indirectlylit) [inline kolibri/plugins/learn/assets/src/vue/breadcrumbs/index.vue]: See @radinamatic recent work here: https://github.com/learningequality/kolibri/pull/355/files

Those used text characters, so maybe it's all right now that we're using SVGs? If so I imagine it needs a `role="presentation"` attribute.

Also, they can they be externalized using `<svg src="..."></svg>`?

File: kolibri/plugins/learn/assets/src/vue/breadcrumbs/index.vue [hunk 3]
```diff
@@ -1,38 +1,92 @@
...
+          Back
+          </span>
+        </a>
+      </span>
+      <span class="parent full-breadcrumbs" v-if="pageMode === $options.PageModes.EXPLORE" v-for="crumb in crumbs">
+        <a v-link="crumbLink(crumb.id)">{{ crumb.title }}</a>
+      </span>
+      <span class="back" v-if="!isRoot && pageMode === $options.PageModes.EXPLORE && pageName !== $options.PageNames.EXPLORE_CONTENT">
```
Comment (indirectlylit) [inline kolibri/plugins/learn/assets/src/vue/breadcrumbs/index.vue]: All of the inline `v-if` statements make it somewhat difficult to understand what is supposed to show and when.

Two options for cleaning up the logic:
- move conditions to computed properties, so that the if statements read like simple sentences
- If some of these if statements are mutually exclusive, consider moving the individual breadcrumb styles to sub-components and using a `<component is="...">` strategy, similar to on the root index.vue pages.

File: kolibri/plugins/learn/assets/src/vue/breadcrumbs/index.vue [hunk 4]
```diff
@@ -1,38 +1,92 @@
...
   const PageNames = require('../../state/constants').PageNames;
+  const constants = require('../../state/constants');
+  const getters = require('../../state/getters');
 
   module.exports = {
-    props: {
-      rootid: {
-        type: String,
-        required: true,
+    mixins: [constants], // makes constants available in $options
```
Comment (indirectlylit) [inline kolibri/plugins/learn/assets/src/vue/breadcrumbs/index.vue]: I've been trying to avoid this pattern because it's confusing. Consider computed properties as mentioned above - then you can reference the constants much more simply.

File: kolibri/plugins/learn/assets/src/vue/breadcrumbs/index.vue [hunk 5]
```diff
@@ -1,38 +1,92 @@
...
-    },
-    computed: {
-      rootLink() {
+      parentLink() {
+        let bread;
+        let id;
+        if (this.pageState.content) {
+          bread = this.pageState.content.breadcrumbs;
+          id = bread[bread.length - 1].id;
+        }
```
Comment (indirectlylit) [inline kolibri/plugins/learn/assets/src/vue/breadcrumbs/index.vue]: all the code above is unused

File: kolibri/plugins/learn/assets/src/vue/breadcrumbs/index.vue [hunk 6]
```diff
@@ -1,38 +1,92 @@
...
+        return {
+          name: PageNames.EXPLORE_TOPIC,
+          params: { id },
+        };
+      },
+      backLink() {
+        let bread;
+        let id;
+        if (this.pageState.topic) {
+          bread = this.pageState.topic.breadcrumbs;
```
Comment (indirectlylit) [inline kolibri/plugins/learn/assets/src/vue/breadcrumbs/index.vue]: what is the point of the bread variable? it's never set to anything except `this.pageState.topic.breadcrumbs`. At the very least it should be `const`

File: kolibri/plugins/learn/assets/src/vue/breadcrumbs/index.vue [hunk 7]
```diff
@@ -1,38 +1,92 @@
...
+          if (bread[bread.length - 1]) {
+            id = bread[bread.length - 1].id;
```
Comment (indirectlylit) [inline kolibri/plugins/learn/assets/src/vue/breadcrumbs/index.vue]: similarly, the `id` never changes - it shouldn't be a `let`

File: kolibri/plugins/learn/assets/src/vue/index.vue [hunk 8]
```diff
@@ -27,10 +27,13 @@
   const PageModes = constants.PageModes;
   const getters = require('../state/getters');
   const store = require('../state/store');
+  require('vue-scroll');
```
Comment (indirectlylit) [inline kolibri/plugins/learn/assets/src/vue/index.vue]: does this do anything?

File: kolibri/plugins/learn/assets/src/vue/index.vue [hunk 9]
```diff
@@ -78,13 +81,18 @@
...
-    right: 2rem
-    z-index: 1
+  .main
+    position: fixed // must be fixed for ie10
+    overflow-y: scroll
+    height: 100%
+    width: 100%
+    padding-left: $left-margin
+    padding-right: $right-margin
+    padding-bottom: 50px
```
Comment (indirectlylit) [inline kolibri/plugins/learn/assets/src/vue/index.vue]: Why was this added? should be handled in `core-base`

File: kolibri/plugins/learn/assets/src/vue/breadcrumbs/index.vue [hunk 10]
```diff
@@ -44,23 +98,58 @@
...
+    vertical-align: middle
+    fill: $core-text-annotation
+    margin-bottom: 2px
+  .full-breadcrumbs
+    @media screen and (max-width: $portrait-breakpoint)
+      display: none
+  .back
+    display: none
+    @media screen and (max-width: $portrait-breakpoint)
+      display: initial
```
Comment (indirectlylit) [inline kolibri/plugins/learn/assets/src/vue/breadcrumbs/index.vue:132]: please add endlines between classes for consistency with the rest of our code base

**Comment by indirectlylit** [comment]
covering recently-added channel switcher:

image

Also seems like maybe a lot of extra space above the search bar?

image

**Comment by indirectlylit** [comment]
This behavior is very cool - but the implementation needs some cleanup. Please see my comments inline.

It also has some conflicts with master - I made a PR for resolving them here: https://github.com/whitzhu/kolibri/pull/10

thank you!

File: kolibri/plugins/learn/assets/src/vue/breadcrumbs/index.vue (hunk 5 above)
Comment (whitzhu) [inline kolibri/plugins/learn/assets/src/vue/breadcrumbs/index.vue]: it is used in the template
https://github.com/learningequality/kolibri/pull/358/files#diff-679db268b88b98c38a2b0e229b39cb6bR20

File: kolibri/plugins/learn/assets/src/vue/index.vue [hunk 11]
```diff
@@ -40,8 +40,6 @@
   const PageModes = constants.PageModes;
   const getters = require('../state/getters');
   const store = require('../state/store');
-  require('vue-scroll');
-
```
Comment (radinamatic) [inline kolibri/plugins/learn/assets/src/vue/index.vue:30]: @whitzhu: Since you are already editing this file, change the line 8 up at the beginning 

from
`<label for="chan-select" class="visuallyhidden">Filter User Type</label>`
to 
`<label for="chan-select" class="visuallyhidden">Select Channel</label>`

Somebody was probably too eager at copy/pasting... :stuck_out_tongue:

**Comment by radinamatic** [comment]
I'm guessing this is still WIP, as there are two Search buttons and two H1 headings as of now:

FF

kolibri dev server - mozilla firefox_632


Chrome:

selection_633

File: kolibri/plugins/learn/assets/src/vue/index.vue [hunk 12]
```diff
@@ -2,6 +2,7 @@
 
   <core-base>
     <main-nav slot="nav"></main-nav>
+    <toolbar slot="above"></toolbar>
     <div slot="above" class="top-wrapper">
       <search-button class='search-btn'></search-button>
       <label for="chan-select" class="visuallyhidden">Filter User Type</label>
```
Comment (radinamatic) [inline kolibri/plugins/learn/assets/src/vue/index.vue]: Label is for "Switch Channels":
`<label for="chan-select" class="visuallyhidden">Switch Channels</label>`

**Comment by radinamatic** [comment]
Another thing I see unnecessary is to truncate the breadcrumbs when there's more than enough space available, even on 760px wide viewport, and especially on 1200px (they are still truncated):

kolibri dev server - mozilla firefox_634

File: kolibri/plugins/learn/assets/src/vue/breadcrumbs/index.vue (hunk 5 above)
Comment (indirectlylit) [inline kolibri/plugins/learn/assets/src/vue/breadcrumbs/index.vue]: Sorry that was bad feedback. What I meant is if you look at the whole function, `id` is left potentially undefined:

``` javascript
      parentLink() {
        let bread;
        let id;
        if (this.pageState.content) {
          bread = this.pageState.content.breadcrumbs;
          id = bread[bread.length - 1].id;
        }
        return {
          name: PageNames.EXPLORE_TOPIC,
          params: { id },
        };
      },
```

I'm guessing what was intended was to create a link back to a different page (e.g. `EXPLORE_CHANNEL`) in certain cases?

A few other notes:
- after @christianmemije's work on channel switching, you need to include the channel ID in the links
- you shouldn't be checking for the existence of the `content` on `pageState` to drive this logic - that will easily break if someone changes how the `pageState` is structured. Instead, check the value of `state.pageName`.
- "bread" as a variable or a component is not very informative :)

File: kolibri/plugins/learn/assets/src/vue/breadcrumbs/index.vue [hunk 13]
```diff
@@ -1,39 +1,79 @@
...
+      <span class="learn-bread" v-if="pageName === allPageNames.LEARN_CONTENT">
+        <first-bread :breadlink="learnRoot" breadtext="Learn"></first-bread>
+      </span>
+
+      <span class="explore-bread" v-if="!isRoot && pageName === allPageNames.EXPLORE_CHANNEL">
+        <first-bread :showarrow='false' :breadlink="exploreRoot" breadtext="Explore"></first-bread>
+      </span>
+
+      <span class="portrait-only" v-if="!isRoot && pageName === allPageNames.EXPLORE_CHANNEL">
+        <first-bread :breadlink="PortraitOnlyParentLink"></first-bread>
```
Comment (indirectlylit) [inline kolibri/plugins/learn/assets/src/vue/breadcrumbs/index.vue]: `PortraitOnlyParentLink` shouldn't be capitalized

File: kolibri/plugins/learn/assets/src/vue/breadcrumbs/index.vue [hunk 14]
```diff
@@ -1,39 +1,79 @@
...
-      rootid: {
-        type: String,
-        required: true,
-      },
-      crumbs: {
-        type: Array,
-        required: true,
-      },
+    components: {
+      'first-bread': require('./back-bread'),
```
Comment (indirectlylit) [inline kolibri/plugins/learn/assets/src/vue/breadcrumbs/index.vue]: inconsistent naming

File: kolibri/plugins/video_mp4_render/assets/src/vue/index.vue [hunk 15]
```diff
@@ -212,10 +223,19 @@
...
     height: 100%
     background-color: #000
 
+  .videowrapperwrapper
+    width: 100%
+    height: 100%
+    background-color: rgba(0, 0, 0, 0)
+    position: relative
+
   .video-js .vjs-menu
```
Comment (indirectlylit) [inline kolibri/plugins/video_mp4_render/assets/src/vue/index.vue:271]: Were these video changes added to this PR by accident? seems unrelated.

Comment (indirectlylit) [inline kolibri/plugins/video_mp4_render/assets/src/vue/index.vue:271]: (You may need to sync with master?)

File: kolibri/plugins/learn/assets/src/vue/breadcrumbs/index.vue (hunk 5 above)
Comment (66eli77) [inline kolibri/plugins/learn/assets/src/vue/breadcrumbs/index.vue]: it seems both pageState and pageName are vuex states, wonder what makes pageName more reliable than pageState?

File: kolibri/plugins/learn/assets/src/vue/breadcrumbs/index.vue [hunk 16]
```diff
@@ -1,39 +1,79 @@
...
     },
     computed: {
-      rootLink() {
-        return { name: PageNames.EXPLORE_CHANNEL };
+      allPageMode() {
+        return PageModes;
+      },
+      allPageNames() {
+        return PageNames;
```
Comment (indirectlylit) [inline kolibri/plugins/learn/assets/src/vue/breadcrumbs/index.vue:46]: I like this pattern - better than `$options`

File: kolibri/plugins/learn/assets/src/vue/breadcrumbs/index.vue (hunk 5 above)
Comment (66eli77) [inline kolibri/plugins/learn/assets/src/vue/breadcrumbs/index.vue]: what if someone changes how the pageName is structured? also the breadcrumbs property is stored inside pageState, that's where the id came from.

Comment (indirectlylit) [inline kolibri/plugins/learn/assets/src/vue/breadcrumbs/index.vue]: pageNames are stored in constants.js and used for routing - they're the single-source-of-truth for which page we're on.

The `id` is undefined if `this.pageState.content` is false

**Comment by indirectlylit** [comment]
note for later: https://trello.com/c/3g4XLseR/405-bug-content-channel-switcher-and-search-functionality-should-be-hidden-when-no-channels-are-available

we can address in a second PR

File: kolibri/plugins/learn/assets/src/vue/breadcrumbs/index.vue (hunk 5 above)
Comment (whitzhu) [inline kolibri/plugins/learn/assets/src/vue/breadcrumbs/index.vue]: if this.pageState is not reliable, is there any way I can get the breadcrumbs? the whole feature relies on this

**Comment by christianmemije** [comment]
@indirectlylit Scroll event doesn't fire as often, used 75 ms as you suggested, but a slightly bigger number seems to also work fine.

File: kolibri/plugins/learn/assets/src/vue/toolbar/index.vue [hunk 17]
```diff
@@ -0,0 +1,238 @@
+<template>
+
+  <div v-bind:class="['toolbar-show', displayToolBar ? 'toolbar-hide' : '' ]" >
```
Comment (radinamatic) [inline kolibri/plugins/learn/assets/src/vue/toolbar/index.vue:3]: I know you are still working on this, just wanted to mention that you should make sure the visual order of elements in the toolbar corresponds to the tabbing order (which is dependent on their position in the code). In current master, for example, channel switcher drop-down `select` is below the search button component in the code, yet the visual order is the other way around, which can cause confusion if user navigates by keyboard. 

So if the visual order in the toolbar is `breadcrumbs - channel select - search button`, make sure that focus by TAB key follows the same order.

Comment (whitzhu) [inline kolibri/plugins/learn/assets/src/vue/toolbar/index.vue:3]: @radinamatic The code within this PR addresses that problem by having
breadcrumb --> channel switch --> search 
you can find it in the `toolbar index.vue`

screen shot 2016-08-15 at 8 56 52 am

Comment (radinamatic) [inline kolibri/plugins/learn/assets/src/vue/toolbar/index.vue:3]: @whitzhu Awesome! I saw that the order of components here is different than on master, just making sure that it is a conscious decision and that it mirrors the visual order of the elements in the page. :wink:

File: kolibri/core/assets/src/vue/core-base.vue [hunk 18]
```diff
@@ -4,7 +4,7 @@
     <nav-bar>
       <slot name="nav"></slot>
     </nav-bar>
-    <div class='main-wrapper' v-scroll='onScroll'>
+    <div class='main-wrapper' v-scroll='scrolled'>
```
Comment (indirectlylit) [inline kolibri/core/assets/src/vue/core-base.vue]: nice

**Comment by indirectlylit** [comment]
@christianmemije - looks good, thank you! 

Minor commentary - no need to refactor: It may have made a little more sense to throttle on the listener side (`toolbar`) rather on the emitter (`core-base`), because the desired rate really depends on what the event is being used for.

**Comment by rtibbles** [comment]
Note that this PR has merge conflicts, and could potentially suffer more if my string wrapping stuff is merged before this.

File: kolibri/plugins/learn/assets/src/vue/toolbar/index.vue [hunk 19]
```diff
@@ -0,0 +1,238 @@
...
+    <breadcrumbs class="breadcrumbs"></breadcrumbs>
+    <div :class="{ 'toggle-menu-on' : more }">
+      <label for="chan-select" :class="[ more ? 'lable-on' : 'visuallyhidden' ]" >Switch Channels</label>
+      <select
+        class="chan-select"
+        :class="[ more ? 'chan-select-mobile-location' : 'chan-select-location' ]"
+        id="chan-select"
+        name="chan-select"
+        v-model="getCurrentChannel"
```
Comment (rtibbles) [inline kolibri/plugins/learn/assets/src/vue/toolbar/index.vue:12]: Unfortunately, this has already been merged, but you should probably avoid `v-model` here, as that does two way data binding.

As the computed property you are using does not have a setter, `v-model` is not appropriate, however, if you wanted, you could add a setter to the computed property, to trigger the action that will update the vuex store.

Comment (rtibbles) [inline kolibri/plugins/learn/assets/src/vue/toolbar/index.vue:12]: @indirectlylit I think we will probably need a technical debt card for this kind of thing.

Comment (indirectlylit) [inline kolibri/plugins/learn/assets/src/vue/toolbar/index.vue:12]: Oh I didn't see this! @christianmemije  is this the cause of some breakage?

Comment (christianmemije) [inline kolibri/plugins/learn/assets/src/vue/toolbar/index.vue:12]: @indirectlylit Not quite the source. You can view what the issues were here: https://github.com/learningequality/kolibri/pull/411
